File: 814-binary-tree-pruning/814-binary-tree-pruning.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode pruneTree(TreeNode root) {
        if(root==null) return null;
        root.left = pruneTree(root.left);
        root.right = pruneTree(root.right);
        return (root.val==1 || root.left!=null || root.right!=null) ? root : null ;
    }
}

// class Solution {
//     public TreeNode pruneTree(TreeNode root) {
//         return dfs(root) ? root : null;
//     }
//     boolean dfs(TreeNode root){
//         if(root==null) return false;
//         boolean l = dfs(root.left);
//         boolean r = dfs(root.right);
//         if(!l) root.left=null;
//         if(!r) root.right=null;
//         return l || r || (root.val==1)?true:false;
//     }
// }

File: 814-binary-tree-pruning/814-binary-tree-pruning.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        root.left = self.pruneTree(root.left);
        root.right = self.pruneTree(root.right);
        return root if (root.val==1 or root.left!=None or root.right!=None) else None

#######################################################################
# class Solution:
#     def pruneTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         def dfs(root):
#             if not root:
#                 return False
#             l=dfs(root.left)
#             r=dfs(root.right)
#             if l==False:
#                 root.left=None
#             if r==False:
#                 root.right=None
#             return l or r or root.val
        
#         return None if dfs(root)==False else root